README
NameTag 2
=========

This is a named entity recognition tool NameTag 2.

It is a public release of the following publication:

Jana Straková, Milan Straka, Jan Hajič (2019): Neural Architectures for Nested
NER through Linearization. In: Proceedings of the 57th Annual Meeting of the
Association for Computational Linguistics, pp. 5326-5331, Association for
Computational Linguistics, Stroudsburg, PA, USA, ISBN 978-1-950737-48-2
(https://aclweb.org/anthology/papers/P/P19/P19-1527/)

NameTag 2 can be used either as a commandline tool (this repository) or by
requesting NameTag webservice (http://lindat.mff.cuni.cz/services/nametag). You
can also run your own NameTag server.

Homepage: https://ufal.mff.cuni.cz/nametag
Web service and demo: http://lindat.mff.cuni.cz/services/nametag/
Contact: strakova@ufal.mff.cuni.cz

1. License
----------

Copyright 2021 Institute of Formal and Applied Linguistics, Faculty of
Mathematics and Physics, Charles University, Czech Republic.

This Source Code Form is subject to the terms of the Mozilla Public License, v.
2.0. If a copy of the MPL was not distributed with this file, You can obtain
one at http://mozilla.org/MPL/2.0/.

2. Please cite as:
------------------

@inproceedings{strakova-etal-2019-neural,
    title = {{Neural Architectures for Nested {NER} through Linearization}},
    author = {Jana Strakov{\'a} and Milan Straka and Jan Haji\v{c}},
    booktitle = {Proceedings of the 57th Annual Meeting of the Association for Computational Linguistics},
    month = jul,
    year = {2019},
    address = {Florence, Italy},
    publisher = {Association for Computational Linguistics},
    url = {https://www.aclweb.org/anthology/P19-1527},
    pages = {5326--5331},
}

3. Requirements
---------------

The software has been developed and tested on Linux. You'll need a machine with
AVX instructions for serving BERT embeddings (see below).

4. Installation of the commandline NameTag 2
--------------------------------------------

4.1 Clone the NameTag 2 GIT repository:

$ git clone https://github.com/ufal/nametag -b nametag2

The main NameTag 2 tagger is called nametag2.py. Running nametag2.py for NER
prediction is described in Section 5. Training your own custom NER model is
described in Section 6.

4.2 Get TensorFlow 1.12

You will need Python virtual environment.

4.2.1 TensorFlow for CPU (locally or SGE)

Create a Python virtual environment with TensorFlow 1.12 called
venv-tf-1.12-cpu in the root of this directory:
 
$ python3 -m venv venv-tf-1.12-cpu
$ venv-tf-1.12-cpu/bin/pip3 install -r requirements.txt

The nametag2.py script is then called using the Python installed in your virtual
environment, which now contains also the appropriate version of TensorFlow.

4.2.2 TensorFlow for GPU

If you plan to scale up for larger experiments or greater speed, you can use
the TensorFlow GPU version. You need to install
Python virtual environment with GPU TensorFlow version.

4.3 Download the NameTag 2 models

Download the latest version of NameTag 2 models from

https://ufal.mff.cuni.cz/nametag/2#models

4.4 Get WEmbedding service

NameTag 2 is using contextualized BERT embeddings computed by Transformers
(https://arxiv.org/abs/1910.03771,
https://github.com/huggingface/transformers).

To this end, we issused a webservice serving transformers BERT embeddings
called WEmbedding service: https://github.com/ufal/wembedding_service.

For on demand BERT embeddings computation, you'll need to setup the WEmbeddings
service on your computer. To setup the WEmbedding service, git clone
https://github.com/ufal/wembedding_service.git as a submodule, then install
Python venv using requirements.txt inside the submodule (see WEmbeddings
repository README). This will allow you to run the WEmbeddings server which
will serve the BERT embeddings to NameTag 2.

5 Running NER prediction with NameTag 2
---------------------------------------

Before you run NameTag 2 for NER prediction, please make sure you have:

  1. installed Python virtual environment (Section 4.2),
  2. downloaded the NameTag 2 models (Section 4.3),
  3. you have a ip:port of a running BERT WEmbeddings service (Section 4.4)

The main NameTag 2 script is called nametag2.py. Example NER prediction usage:

venv-tf-1.12-cpu/bin/python3 nametag2.py --test_data=examples/en_input.conll --predict=models/english-conll-200831 --bert=localhost:8000

Example usage with more documentation on input and output formats can be found
in the nametag2.py script.

6. Training NameTag 2
---------------------

The main NameTag 2 script nametag2.py can be used for training a custom corpus.
It will do so when provided the parameters --train_data. Optionally, --dev_data
can be provided and many other training parameters that control the training.

The input data file format is a vertical file, one token and its label per
line, separated by a tabulator; sentences delimited by newlines (such as
a first and fourth column in a well-known CoNLL-2003 IOB shared task corpus).
An example of such input file can be found in nametag2.py and in
examples/en_gold.conll.

Example usage:

venv-tf-1.12-cpu/bin/python3 nametag2.py --threads=4 --batch_size=4 --train_data=train.conll --dev_data=dev.conll --epochs=10:1e-3,8:1e-4 --save_checkpoint=my_models --bert_train=train_bert.npz --bert_dev=dev_bert.npz

While it is technically possible to compute the BERT embeddings on demand for
each shuffled batch in each step, we strongly recommend to precompute the BERT
embeddings as numpy arrays, one array per each token in your vertical
CoNLL-like file and store these as .npz files. The argument --bert_train and
--bert_dev will load the precomputed BERT embeddings from the npz file and
reuse them during training, speeding up the training process. 

7. NameTag 2 server
-------------------

See nametag2_server.py.

The mandatory arguments are given in this order:

- port
- WEmbeddings server (host:port)
- default model name
- each following triple of arguments defines a model, of which
    - first argument is the model name
    - second argument is the model directory
    - third argument are the acknowledgemets to append

Example server usage:

$ venv/bin/python nametag2_server.py 8001 localhost:8000 czech-cnec2.0-200831 czech-cnec2.0-200831 models/czech-cnec2.0-200831/ ack-text


CHANGES
Version 1.2.1-dev
-----------------


Version 1.2.0 [18 Feb 23]
-------------------------
- Add FormCapitalization feature template.
- Add GazetteersEnhanced feature template, much improving gazetteers
  handling, including better matching, possibility of hard gazetteers
  and support for loading additional gazetteers to an existing model.
- Export recognizable named entities from the recognizer.
- Export gazetteers from the recognizer (only GazetteersEnhanced).
- Add several nametag_server options.
- Update MorphoDita to 1.11.1.
- On Windows, the file paths are now UTF-8 encoded, instead of ANSI.
  This change affects the API, binary arguments, and program outputs.
- Add ARM64 macOS build.
- The Windows binaries are now compiled with VS 2019, older systems
  than Windows 7 are no longer supported.
- The Python wheels are provided for Pythons 3.6-3.11.


Version 1.1.2 [01 Jul 17]
-------------------------
- Allow specifying custom path to C++ library in Java.
- Fix bug causing a memory leak on g++.
- Add --log option to the REST server.


Version 1.1.1 [05 May 16]
-------------------------
- Remove forgotten dllimport attribute (fixes compilation on Windows).


Version 1.1.0 [05 May 16]
-------------------------
- Change license from LGPL to MPL 2.0.
- Fix problem in external tagger, which was unusable.
- Add support for OS X 10.7 and later.
- Support compilation on Visual Studio 2015.
- Add C# bindings.
- Add REST server using MicroRestD http://github.com/ufal/microrestd.
- Start using Semantic Versioning http://semver.org/.
- Remove support for shared library build.
- Use C++ Builtem http://github.com/ufal/cpp_builtem as build system.
- Use embedded version of MorphoDiTa instead of using GIT submodule.
- Use upstream version of C++ Utils http://github.com/ufal/cpp_utils.
- Use upstream version of Unilib http://github.com/ufal/unilib.
- Use PyTypeObject in Python bindings instead of proxy classes.
- Use C++ iostreams instead of C stdio.
- Use t2t_docsys http://github.com/ufal/t2t_docsys as documentation system.


Version 1.0.0 [11 Apr 14]
-------------------------
- Initial public release.


nametag2.py
#!/usr/bin/env python3
# coding=utf-8
#
# Copyright 2021 Institute of Formal and Applied Linguistics, Faculty of
# Mathematics and Physics, Charles University, Czech Republic.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""Nested NER training and prediction script.

Before running this script, please see README on instructions how to setup
your Python virtual environment with TensorFlow and where to get the models
and BERT WEmbeddings server.

Using NameTag 2 for NER prediction
----------------------------------

Before you run NameTag 2 for NER prediction, please make sure you have:

  1. installed Python virtual environment,
  2. downloaded the NameTag 2 models,
  3. you have an ip:port of a running BERT WEmbeddings service (e.g.,
     localhost:8000).

NER prediction example usage:

$ venv-tf-1.12-cpu/bin/python ./nametag2.py --test_data=examples/en_input.conll --predict=models/en_CoNLL_en --bert=localhost:8000

test_data: Input file. The expected format is a segmented and tokenized vertical
           format: one token per line, sentences delimited with newline.
predict: Path to model directory.
bert: ip:port of a running WEmbeddings server.

Input example (also please see examples/en_input.conll):

John
loves
Mary
.

Mary
loves
John.

Output and gold example (also please see examples/en_gold.conll):

John  B-PER
loves O
Mary  B-PER
.

Mary  B-PER
loves O
John  B-PER
.     O

Training NameTag 2
------------------

The main NameTag 2 script nametag2.py can be used for training a custom corpus.
It will do so when provided the parameters --train_data. Optionally, --dev_data
can be provided and many other training parameters that control the training.

The input data file format is a vertical file, one token and its label per
line, separated by a tabulator; sentences delimited by newlines (such as
a first and fourth column in a well-known CoNLL-2003 IOB shared task corpus).
An example of such input file can be found in examples/en_gold.conll.

Example usage:

venv-tf-1.12-cpu/bin/python3 nametag2.py --threads=4 --batch_size=4 --train_data=train.conll --dev_data=dev.conll --epochs=10:1e-3,8:1e-4 --save_checkpoint=my_models --bert_train=train_bert.npz --bert_dev=dev_bert.npz

While it is technically possible to compute the BERT embeddings on demand for
each shuffled batch in each step, we strongly recommend to precompute the BERT
embeddings as numpy arrays, one array per each token in your vertical
CoNLL-like file and store these as .npz files. The argument --bert_train and
--bert_dev will load the precomputed BERT embeddings from the npz file and
reuse them during training, speeding up the training process.
"""


import io
import json
import pickle
import sys

import numpy as np

import nametag2_dataset
import nametag2_network


if __name__ == "__main__":
    import argparse
    import datetime
    import os
    import re

    # Fix random seed for network initialization before training
    np.random.seed(42)

    # Parse arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("--batch_size", default=64, type=int, help="Batch size.")

    # Computes BERT embeddings on demand by requesting WEmbeddings server at ip:port.
    parser.add_argument("--bert", default=None, type=str, nargs="?", const="", help="Get BERT embeddings by requesting ip:port.")

    # For training, precompute BERT embeddings to speed things up.
    parser.add_argument("--bert_dev", default=None, type=str, help="Pretrained BERT embeddings for dev data in file.")
    parser.add_argument("--bert_test", default=None, type=str, help="Pretrained BERT embeddings for test data in file.")
    parser.add_argument("--bert_train", default=None, type=str, help="Pretrained BERT embeddings for train data in file.")

    parser.add_argument("--beta_2", default=0.98, type=float, help="Beta 2.")
    parser.add_argument("--corpus", default="", type=str, help="[english-conll|german-conll|dutch-conll|spanish-conll|czech-cnec2.0]")
    parser.add_argument("--cle_dim", default=128, type=int, help="Character-level embedding dimension.")
    parser.add_argument("--decoding", default="seq2seq", type=str, help="Decoding: [CRF|ME|LSTM|seq2seq].")
    parser.add_argument("--dev_data", default=None, type=str, help="Dev data.")
    parser.add_argument("--dropout", default=0.5, type=float, help="Dropout rate.")

    # E.g.: 10:1e-3,8:1e-4
    parser.add_argument("--epochs", default="10:1e-3", type=str, help="Epochs and learning rates.")

    parser.add_argument("--label_smoothing", default=0, type=float, help="Label smoothing.")
    parser.add_argument("--max_sentences", default=None, type=int, help="Limit number of training sentences (for debugging).")
    parser.add_argument("--max_labels_per_token", default=5, type=int, help="Maximum labels per token.")
    parser.add_argument("--name", default=None, type=str, help="Experiment name.")
    parser.add_argument("--predict", default=None, type=str, help="Predict using the saved checkpoint.")
    parser.add_argument("--rnn_cell", default="LSTM", type=str, help="RNN cell type.")
    parser.add_argument("--rnn_cell_dim", default=256, type=int, help="RNN cell dimension.")
    parser.add_argument("--rnn_layers", default=1, type=int, help="Number of hidden layers.")
    parser.add_argument("--save_checkpoint", default=None, type=str, help="Save checkpoint to path.")
    parser.add_argument("--test_data", default=None, type=str, help="Test data.")
    parser.add_argument("--time", default=False, action="store_true", help="Measure prediction time.")
    parser.add_argument("--train_data", default=None, type=str, help="Training data.")
    parser.add_argument("--threads", default=4, type=int, help="Maximum number of threads to use.")
    parser.add_argument("--we_dim", default=256, type=int, help="Word embedding dimension.")
    parser.add_argument("--word_dropout", default=0.2, type=float, help="Word dropout.")
    args = parser.parse_args()

    if args.predict:
        # Load saved options from the model
        with open("{}/options.json".format(args.predict), mode="r") as options_file:
            args = argparse.Namespace(**json.load(options_file))
        parser.parse_args(namespace=args)
    else:
        # Create TensorFlow logdir name
        logargs = dict(vars(args).items())
        del logargs["bert_dev"]
        del logargs["bert_test"]
        del logargs["bert_train"]
        del logargs["beta_2"]
        del logargs["cle_dim"]
        del logargs["dev_data"]
        del logargs["dropout"]
        del logargs["label_smoothing"]
        del logargs["max_sentences"]
        del logargs["rnn_cell_dim"]
        del logargs["test_data"]
        del logargs["threads"]
        del logargs["train_data"]
        del logargs["we_dim"]
        del logargs["word_dropout"]
        logargs["bert"] = 1 if args.bert_train else 0

        args.logdir = "logs/{}-{}-{}".format(
            os.path.basename(__file__),
            datetime.datetime.now().strftime("%Y-%m-%d_%H%M%S"),
            ",".join(("{}={}".format(re.sub("(.)[^_]*_?", r"\1", key), re.sub("^.*/", "", value) if type(value) == str else value)
                      for key, value in sorted(logargs.items())))
        )
        os.makedirs(args.logdir, exist_ok=True)
       
        # Check if directory exists if saving checkpoint
        if args.save_checkpoint:
            if os.path.isdir(args.save_checkpoint):
                print("Checkpoint will be saved to existing directory: {}".format(args.save_checkpoint), file=sys.stderr, flush=True)
            else:
                print("Path to checkpoint does not exist, creating directory {}".format(args.save_checkpoint), file=sys.stderr, flush=True)
                os.mkdir(args.save_checkpoint)

        # Dump passed options to allow future prediction.
        if args.save_checkpoint:
            with open("{}/options.json".format(args.save_checkpoint), mode="w") as options_file:
                json.dump(vars(args), options_file, sort_keys=True)

    # Postprocess args
    args.epochs = [(int(epochs), float(lr)) for epochs, lr in (epochs_lr.split(":") for epochs_lr in args.epochs.split(","))]

    # Print debug outputs
    print("Number of threads: {}".format(args.threads), file=sys.stderr, flush=True)
    print("Batch size: {}".format(args.batch_size), file=sys.stderr, flush=True)

    # Setup BERT embeddings server in subprocess if empty --bert given
    # (otherwise read from file or use --bert=ip:port for requests).
    if args.bert is not None and not args.bert: # --bert
        print("Please provide ip:port WEmbeddings server for --bert. Empty --bert is not supported.", file=sys.stderr, flush=True)
        sys.exit(1)

    # Load the data
    print("Loading data...", file=sys.stderr, flush=True)
    seq2seq = args.decoding == "seq2seq"
    if not args.predict:
        train = nametag2_dataset.NameTag2Dataset(filename=args.train_data,
                                                 max_sentences=args.max_sentences,
                                                 seq2seq=seq2seq,
                                                 bert_filename=args.bert_train if not args.predict else None,
                                                 bert=args.bert)
        train.save_mappings("{}/mappings.pickle".format(args.save_checkpoint))
        if args.dev_data:
            dev = nametag2_dataset.NameTag2Dataset(filename=args.dev_data,
                                                   train=train,
                                                   shuffle_batches=False,
                                                   seq2seq=seq2seq,
                                                   bert_filename=args.bert_dev,
                                                   bert=args.bert)
    else:
        with open("{}/mappings.pickle".format(args.predict), mode="rb") as mappings_file:
            train = pickle.load(mappings_file)

    if args.test_data:
        test = nametag2_dataset.NameTag2Dataset(filename=args.test_data,
                                                train=train,
                                                shuffle_batches=False,
                                                seq2seq=seq2seq,
                                                bert_filename=args.bert_test,
                                                bert=args.bert)

    # Character-level embeddings
    args.including_charseqs = (args.cle_dim > 0)

    # Construct the network
    network = nametag2_network.Network(threads=args.threads)
    network.construct(args,
                      num_forms=len(train.factors[train.FORMS].words),
                      num_form_chars=len(train.factors[train.FORMS].alphabet),
                      num_tags=len(train.factors[train.TAGS].words),
                      tag_bos=train.factors[train.TAGS].words_map["<bos>"],
                      tag_eow=train.factors[train.TAGS].words_map["<eow>"],
                      bert_dim=train.bert_dim(),
                      predict_only=args.predict)

    # Predict only
    if args.predict:
        network.saver.restore(network.session, "{}/model".format(args.predict.rstrip("/")))
        print("Predicting test data using previously saved checkpoint.", file=sys.stderr)
        output = io.StringIO()
        network.predict("test", test, args, output, evaluating=False)
        output = output.getvalue()
        output = network.postprocess(output)
        sys.stdout.write(output)

    # Train and predict
    else:
        # Train
        for epochs, learning_rate in args.epochs:
            for epoch in range(epochs):
                print("Epoch: {}, learning rate: {}".format(epoch, learning_rate), file=sys.stderr, flush=True)
                network.train_epoch(train, learning_rate, args)
                dev_score = 0
                if args.dev_data:
                    network.evaluate("dev", dev, args)
                    
                    # Undocumented external F1 evaluation option (see network.evaluate)
                    # dev_score = network.evaluate("dev", dev, args)
                    # print("Dev F1: {}".format(dev_score), file=sys.stderr, flush=True)
        # Save network
        if args.save_checkpoint:
            network.saver.save(network.session, "{}/model".format(args.save_checkpoint), write_meta_graph=False)
        # Test
        if args.test_data:
            network.evaluate("test", test, args)
            
            # Undocumented external F1 evaluation option (see network.evaluate)
            # test_score = network.evaluate("test", test, args)
            # print("Test F1: {}".format(test_score), file=sys.stderr, flush=True)
